### PR TITLE
Working progress

### PR DESCRIPTION
**Pull request by veronikaslc**

### CODE REVIEW

File: components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractHGNCSolrOntologyService.java [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.phenotips.ontology.internal.solr;
+
+import org.phenotips.obo2solr.ParameterPreparer;
+import org.phenotips.obo2solr.TermData;
```
Comment (marta-) [inline components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractHGNCSolrOntologyService.java]: You should not create dependencies from this component to the obo2solr package. If some classes from obo2solr are useful here, consider extracting them in another package that is used by both this component and the obo2solr one.

Comment (sdumitriu) [inline components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractHGNCSolrOntologyService.java]: Because obo2solr is deprecated.

File: components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractHGNCSolrOntologyService.java [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+import net.sf.json.JSONObject;
+import net.sf.json.JSONSerializer;
+
+/**
+ * Ontologies processed from HGNC file share much of the processing code.
+ *
+ * @since 1.2RC1
+ * @version $Id$
+ */
+public abstract class AbstractHGNCSolrOntologyService extends AbstractSolrOntologyService
```
Comment (marta-) [inline components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractHGNCSolrOntologyService.java]: Not a good name for the class. It would be if expected multiple implementations HGNC implementations, all extending the AbstractHGNC... class. But that can't be the case, we don't realistically have more options branching from HGNC. Instead, this should be an abstract class that can deal with generic CSV-like files (so AbstractCSVSolrOntologyService, for example).

File: components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractHGNCSolrOntologyService.java [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+    protected static final String ALTERNATIVE_ID_FIELD_NAME = "alt_id";
+
+    protected static final String VERSION_FIELD_NAME = "version";
+
+    protected static final String SIZE_FIELD_NAME = "size";
+    protected static final String ROWS_FIELD_NAME = "rows";
+
+    private static final String FIELD_VALUE_SEPARATOR = "\\t";
+
+    private String infoServiceURL = "http://rest.genenames.org/info";
```
Comment (marta-) [inline components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractHGNCSolrOntologyService.java]: As per my comment above, everything specific to HGNC should move to the class extending this one.

File: components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractHGNCSolrOntologyService.java [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+        try {
+            url = new URL(ontologyUrl);
+        } catch (MalformedURLException ex) {
+            return null;
+        }
+        return transform(url, fieldSelection);
+    }
+
+    private Map<String, TermData> transform(URL url, Map<String, Double> fieldSelection)
+    {
```
Comment (marta-) [inline components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractHGNCSolrOntologyService.java]: Have you considered using https://commons.apache.org/proper/commons-csv/ for parsing the input?

File: components/ontology-access/hgnc/api/src/main/java/org/phenotips/ontology/internal/GeneHGNCNomenclature.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    private static final Pattern ID_PATTERN = Pattern.compile("^HGNC:[0-9]+$", Pattern.CASE_INSENSITIVE);
+
+    @Override
+    public String getDefaultOntologyLocation()
+    {
+        return "http://www.genenames.org/cgi-bin/download?col=gd_hgnc_id&col=gd_app_sym&col=gd_app_name"
+            + "&col=gd_prev_sym&col=gd_aliases&col=gd_pub_acc_ids&col=gd_pub_eg_id&col=gd_pub_ensembl_id&col="
+            + "gd_pub_refseq_ids&col=family.id&col=family.name&col=md_eg_id&col=md_mim_id&col=md_refseq_id&col="
+            + "md_prot_id&col=md_ensembl_id&status=Approved&status_opt=2&where=&order_by=gd_app_sym_sort&format="
+            + "text&hgnc_dbtag=on&submit=submit";
```
Comment (marta-) [inline components/ontology-access/hgnc/api/src/main/java/org/phenotips/ontology/internal/GeneHGNCNomenclature.java]: Most likely the URL looks like this because it's still work in progress. The parameters should of course be defined in static final fields, and the URL constructed from these, in a more maintainable manner.

**Comment by marta-** [comment]
Don't forget to declare the component in components.txt, in order to make it known to the component manager.

File: components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractHGNCSolrOntologyService.java [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+import org.apache.solr.common.SolrDocumentList;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.SolrInputDocument;
+
+import net.sf.json.JSONException;
+import net.sf.json.JSONObject;
+import net.sf.json.JSONSerializer;
+
+/**
+ * Ontologies processed from HGNC file share much of the processing code.
```
Comment (sdumitriu) [inline components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractHGNCSolrOntologyService.java]: HGNC is an ontology, there are no other ontologies from HGNC.

File: components/export/api/src/checkstyle/checkstyle-suppressions.xml [hunk 7]
```diff
@@ -24,4 +24,6 @@
 <suppressions>
   <suppress checks=".*" files="DataToCellConverter.java"/>
   <suppress checks="CyclomaticComplexity" files="Styler.java" lines="155, 209"/>
+  <suppress checks="ClassDataAbstractionCoupling" files="GeneHGNCNomenclature.java" lines="77, 341"/>
```
Comment (sdumitriu) [inline components/export/api/src/checkstyle/checkstyle-suppressions.xml]: These are in the wrong file. Each module can have its own `checkstyle-suppressions.xml` file, so you must add a new file in the `components/ontology-access/hgnc/` module and [enable it](https://github.com/phenotips/phenotips/blob/master/components/export/api/pom.xml#L33) in the pom.

File: components/obo2solr/src/main/java/org/phenotips/obo2solr/TermData.java [hunk 8]
```diff
@@ -45,7 +45,7 @@ public void clear()
         super.clear();
     };
 
-    String getId()
+    public String getId()
```
Comment (sdumitriu) [inline components/obo2solr/src/main/java/org/phenotips/obo2solr/TermData.java]: Is this change still needed?

File: components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractOBOSolrOntologyService.java [hunk 9]
```diff
@@ -174,8 +174,10 @@ public String getVersion()
                 firstDoc = termList.get(0);
                 return firstDoc.getFieldValue(VERSION_FIELD_NAME).toString();
             }
-        } catch (SolrServerException | SolrException | IOException ex) {
+        } catch (SolrServerException | SolrException ex) {
             this.logger.warn("Failed to query ontology version: {}", ex.getMessage());
+        } catch (IOException ex) {
+            this.logger.error("IOException while getting ontology version", ex);
```
Comment (sdumitriu) [inline components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractOBOSolrOntologyService.java]: Why did you split this?

File: components/ontology-access/hgnc/api/src/main/resources/META-INF/components.txt [hunk 10]
```diff
@@ -1 +1 @@
-org.phenotips.ontology.internal.GeneNomenclature
```
Comment (sdumitriu) [inline components/ontology-access/hgnc/api/src/main/resources/META-INF/components.txt]: You shouldn't disable the old one.

File: components/obo2solr/src/main/java/org/phenotips/obo2solr/TermData.java (hunk 8 above)
Comment (veronikaslc) [inline components/obo2solr/src/main/java/org/phenotips/obo2solr/TermData.java]: Nope, actually I don't need this anymore. Will correct in next commit.

File: components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractOBOSolrOntologyService.java (hunk 9 above)
Comment (veronikaslc) [inline components/ontology-access/api/src/main/java/org/phenotips/ontology/internal/solr/AbstractOBOSolrOntologyService.java]: I guess due to multiple code changes during development. Will combine now.

**Comment by sdumitriu** [comment]
There should be some Solr configuration files as well (schema.xml, solrconfig.xml).

**Comment by veronikaslc** [comment]
Added checkstyle suppressions to ontology-access\hgnc\api\src\checkstyle, but still have checkstyle errors

**Comment by sdumitriu** [comment]
You only suppressed some checks, but you still have other errors which should be solved. You can see the exact problems if you scroll up in the console. For example: https://ci.phenotips.org/job/PhenoTips-pull-requests/421/org.phenotips$ontology-hgnc-access-api/console

**Comment by danielpgross** [comment]
Is there a better way to work with these branches so that irrelevant commits and changes don't show here? See below, for example:
working_progress_by_veronikaslc_ _pull_request__1657_ _phenotips_phenotips

**Comment by mjshepherd** [comment]
Yes there is...rebase

File: components/vocabularies/api/src/main/java/org/phenotips/vocabulary/internal/solr/AbstractSolrVocabulary.java [hunk 11]
```diff
@@ -56,6 +56,13 @@
     /** The name of the ID field. */
     protected static final String ID_FIELD_NAME = "id";
 
+    // to "boost" the score of documents in cases where all of the terms in the "q" param appear in close proximity
+    protected static final String COMMON_PARAMS_PF = "pf";
```
Comment (sdumitriu) [inline components/vocabularies/api/src/main/java/org/phenotips/vocabulary/internal/solr/AbstractSolrVocabulary.java]: You should use `DisMaxParams.PF` and `DisMaxParams.QF` instead of redefining them here.

**Comment by sdumitriu** [comment]
Manually merged into master.
